File: packages/geosolve-collaboration/src/semantic.ts
// SPDX-License-Identifier: GPL-3.0-or-later
import { counter, unicode, encode, decode } from "./host-codec.js";
import type { OperationId } from "./host.js";

/** Canonical names and dependencies come from the trusted compiler, never scene labels. */
export interface SemanticTarget { readonly object: string; readonly generation: number }
export interface DeletionPlan { readonly roots: readonly SemanticTarget[]; readonly closure: readonly SemanticTarget[] }
export interface PropertyAddress { readonly target: SemanticTarget; readonly property: string }
export interface StatementPosition { readonly previous: SemanticTarget | null; readonly next: SemanticTarget | null }
export interface ObjectDescription {
  readonly target: SemanticTarget;
  readonly dependencies: readonly SemanticTarget[];
  readonly payload: unknown;
  readonly position: StatementPosition;
}
export interface DependencyChange { readonly target: SemanticTarget; readonly before: readonly SemanticTarget[]; readonly after: readonly SemanticTarget[] }
export interface ReorderChange { readonly target: SemanticTarget; readonly before: StatementPosition; readonly after: StatementPosition }
export interface StructuralInverse {
  readonly create: readonly ObjectDescription[];
  readonly delete: DeletionPlan | null;
  readonly dependencies: readonly DependencyChange[];
  readonly reorders: readonly ReorderChange[];
  readonly restored: readonly { readonly before: SemanticTarget; readonly after: SemanticTarget }[];
}
export interface CreationPosition {
  readonly previous: SemanticTarget | SemanticTargetReference | null;
  readonly next: SemanticTarget | SemanticTargetReference | null;
}
export interface StructuralRecord {
  readonly created?: readonly { readonly object: string; readonly payload: unknown; readonly position: CreationPosition }[];
  readonly deleted?: readonly { readonly target: SemanticTarget; readonly payload: unknown; readonly position: StatementPosition }[];
  readonly reorders?: readonly ReorderChange[];
}
export interface PropertyChange { readonly address: PropertyAddress; readonly before: unknown; readonly after: unknown }
export interface SemanticHostConfiguration {
  readonly documentEpoch: string;
  /** Fresh process epoch after durable reconstruction. */
  readonly serverEpoch: string;
  readonly objects?: readonly { readonly object: string; readonly dependencies?: readonly string[] }[];
  readonly limits?: {
    readonly maxObjectsIncludingTombstones?: number;
    readonly maxDependenciesPerObject?: number;
    readonly maxTotalDependencies?: number;
  };
}
/** Store both native JSON strings unchanged with the authenticated model revision. */
export interface SemanticCheckpoint {
  readonly documentEpoch: string;
  readonly revision: number;
  readonly targetsJson: string;
  readonly historyJson: string;
}
export interface SemanticSnapshot {
  readonly revision: number;
  readonly historyRevision: number;
  readonly highWater: number;
  readonly dependencyCount: number;
  readonly hasPendingStage: boolean;
  readonly needsRecovery: boolean;
}
/** Descriptor only; the checked InversePlan remains owned by native Rust. */
export interface PreparedSemanticInverse {
  readonly ticket: string;
  readonly basisRevision: number;
  readonly contribution: OperationId;
  readonly direction: "undo" | "redo";
  readonly structural: StructuralInverse;
  readonly changes: readonly PropertyChange[];
}
export type SemanticTargetReference = { readonly kind: "existing"; readonly target: SemanticTarget }
  | { readonly kind: "created"; readonly object: string };
export interface SemanticRecord {
  readonly basisRevision: number;
  readonly revision: number;
  readonly operation: OperationId;
  readonly changes: readonly PropertyChange[];
}
/** Validated compiler inventory updates. Deletions retain exactly reviewed closures;
 * all created names are allocated before forward dependency references are resolved.
 * Supply record.structural with trusted compiler observations to put lifecycle,
 * dependency and reorder contributions in the same per-user Undo timeline.
 */
export interface SemanticTransaction {
  readonly basisRevision: number;
  readonly revision: number;
  readonly create?: readonly { readonly object: string; readonly dependencies?: readonly SemanticTargetReference[] }[];
  readonly deletions?: readonly DeletionPlan[];
  readonly dependencies?: readonly { readonly target: SemanticTargetReference; readonly dependencies: readonly SemanticTargetReference[] }[];
  readonly record?: { readonly operation: OperationId; readonly changes: readonly PropertyChange[]; readonly structural?: StructuralRecord };
}
/** Persist exact strings with accepted model/source/authority outcome before commit. */
export interface SemanticStage {
  readonly status: "staged";
  readonly stageId: string;
  readonly basisRevision: number;
  readonly revision: number;
  readonly targetsJson: string;
  readonly historyJson: string;
  readonly created: readonly SemanticTarget[];
}
export type PersistSemanticStage = (stage: SemanticStage) => Promise<void>;
export interface SemanticNativeHandle {
  snapshot(): string;
  checkpoint(): string;
  current(object: string): string;
  authenticate(targetJson: string): void;
  planDelete(rootsJson: string): string;
  authenticateDelete(planJson: string): void;
  propertyOwner(addressJson: string): string;
  prepareUndo(userId: string): string;
  prepareRedo(userId: string): string;
  release(ticket: string): void;
  stageValidatedRecord(requestJson: string): string;
  stageValidatedTransaction(requestJson: string): string;
  stageValidatedInverse(ticket: string, operationJson: string, revisionJson: string): string;
  commitStage(stageId: string): string;
  failStage(stageId: string): void;
  discardUnpersistedStage(stageId: string): void;
  free(): void;
}
export interface SemanticWasmModule {
  default(options: { module_or_path: Uint8Array | URL | Response | WebAssembly.Module }): Promise<unknown>;
  TrustedSemanticHost: {
    new(configurationJson: string): SemanticNativeHandle;
    restore(configurationJson: string, checkpointJson: string): SemanticNativeHandle;
  };
}
export interface SemanticHostOptions {
  readonly configuration: SemanticHostConfiguration;
  readonly checkpoint?: SemanticCheckpoint;
  readonly wasmModule?: SemanticWasmModule;
  readonly wasm?: Uint8Array | URL | Response | WebAssembly.Module;
}
/** Trusted Node host adapter over actual Rust TargetLedger + ContributionHistory.
 * Synchronous staging stays separate from asynchronous durable publication.
 */
export class TrustedSemanticHost {
  private disposed = false;
  private pending?: SemanticStage;
  private readonly prepared = new Set<PreparedSemanticInverse>();
  constructor(private readonly native: SemanticNativeHandle) {}
  snapshot(): SemanticSnapshot { this.live(); return decode(this.native.snapshot()); }
  checkpoint(): SemanticCheckpoint { this.live(); return decode(this.native.checkpoint()); }
  current(object: string): SemanticTarget | null { this.live(); unicode(object); return decode(this.native.current(object)); }
  authenticate(target: SemanticTarget): void { this.live(); counter(target.generation); this.native.authenticate(encode(target)); }
  planDelete(roots: readonly SemanticTarget[]): DeletionPlan { this.live(); return decode(this.native.planDelete(encode(roots))); }
  authenticateDelete(plan: DeletionPlan): void { this.live(); this.native.authenticateDelete(encode(plan)); }
  propertyOwner(address: PropertyAddress): OperationId | null { this.live(); return decode(this.native.propertyOwner(encode(address))); }
  prepareUndo(userId: string): PreparedSemanticInverse {
    this.live(); unicode(userId); return this.retain(this.native.prepareUndo(userId));
  }
  prepareRedo(userId: string): PreparedSemanticInverse {
    this.live(); unicode(userId); return this.retain(this.native.prepareRedo(userId));
  }
  release(prepared: PreparedSemanticInverse): void {
    this.ownsPrepared(prepared); this.native.release(prepared.ticket); this.prepared.delete(prepared);
  }
  stageValidatedRecord(request: SemanticRecord): SemanticStage {
    this.live(); counter(request.basisRevision); counter(request.revision);
    return this.stage(this.native.stageValidatedRecord(encode(request)));
  }
  stageValidatedTransaction(request: SemanticTransaction): SemanticStage {
    this.live(); counter(request.basisRevision); counter(request.revision);
    return this.stage(this.native.stageValidatedTransaction(encode(request)));
  }
  stageValidatedInverse(prepared: PreparedSemanticInverse, operation: OperationId, revision: number): SemanticStage {
    this.ownsPrepared(prepared); counter(revision);
    return this.stage(this.native.stageValidatedInverse(prepared.ticket, encode(operation), encode(revision)));
  }
  /** Call only after the exact stage/model/source/authority envelope is durable. */
  commitStage(stage: SemanticStage): SemanticSnapshot {
    this.owns(stage);
    const result = decode<SemanticSnapshot>(this.native.commitStage(stage.stageId));
    this.pending = undefined; this.prepared.clear(); return result;
  }
  /** Only before any persistence has begun; uncertain writes require failStage. */
  discardUnpersistedStage(stage: SemanticStage): void {
    this.owns(stage); this.native.discardUnpersistedStage(stage.stageId); this.pending = undefined;
  }
  failStage(stage: SemanticStage): void {
    this.owns(stage); this.native.failStage(stage.stageId); this.pending = undefined; this.prepared.clear();
  }
  async record(request: SemanticRecord, persist: PersistSemanticStage): Promise<SemanticSnapshot> {
    return this.persist(this.stageValidatedRecord(request), persist);
  }
  async transact(request: SemanticTransaction, persist: PersistSemanticStage): Promise<SemanticSnapshot> {
    return this.persist(this.stageValidatedTransaction(request), persist);
  }
  async inverse(prepared: PreparedSemanticInverse, operation: OperationId, revision: number, persist: PersistSemanticStage): Promise<SemanticSnapshot> {
    return this.persist(this.stageValidatedInverse(prepared, operation, revision), persist);
  }
  dispose(): void {
    if (!this.disposed) {
      if (this.pending) throw Error("Resolve pending semantic persistence before disposing the host");
      this.disposed = true; this.prepared.clear(); this.native.free();
    }
  }
  private retain(text: string): PreparedSemanticInverse {
    const result = decode<PreparedSemanticInverse>(text); this.prepared.add(result); return result;
  }
  private stage(text: string): SemanticStage {
    const result = decode<SemanticStage>(text); this.pending = result; return result;
  }
  private owns(stage: SemanticStage): void {
    this.live(); if (stage !== this.pending) throw Error("Unknown or stale semantic stage");
  }
  private ownsPrepared(prepared: PreparedSemanticInverse): void {
    this.live(); if (!this.prepared.has(prepared)) throw Error("Unknown or stale semantic inverse ticket");
  }
  private async persist(stage: SemanticStage, persist: PersistSemanticStage): Promise<SemanticSnapshot> {
    try { await persist(stage); }
    catch (failure) { this.failStage(stage); throw failure; }
    return this.commitStage(stage);
  }
  private live(): void { if (this.disposed) throw Error("Semantic host has been disposed"); }
}

export async function createTrustedSemanticHost(options: SemanticHostOptions): Promise<TrustedSemanticHost> {
  const configuration = encode(options.configuration);
  if (options.checkpoint) counter(options.checkpoint.revision);
  const checkpoint = options.checkpoint === undefined ? undefined : encode(options.checkpoint);
  const path = "./wasm/geosolve_collaboration_wasm.js";
  const module: SemanticWasmModule = options.wasmModule ?? await import(path);
  let bytes = options.wasm;
  if (!bytes) {
    const url = new URL("./wasm/geosolve_collaboration_wasm_bg.wasm", import.meta.url);
    if (url.protocol === "file:") {
      const nodeFs = "node:fs/promises";
      bytes = await (await import(nodeFs)).readFile(url);
    } else bytes = url;
  }
  await module.default({ module_or_path: bytes! });
  return new TrustedSemanticHost(checkpoint === undefined
    ? new module.TrustedSemanticHost(configuration)
    : module.TrustedSemanticHost.restore(configuration, checkpoint));
}

File: crates/geosolve-collaboration-wasm/tests/semantic.rs
// SPDX-License-Identifier: GPL-3.0-or-later
use geosolve_collaboration_wasm::TrustedSemanticHost;
use serde_json::{Value, json};

fn config(process: &str) -> String {
    json!({"documentEpoch":"doc-1","serverEpoch":process,"objects":[{"object":"edge","dependencies":["width"]},{"object":"width"},{"object":"height"}]}).to_string()
}
fn host() -> TrustedSemanticHost {
    TrustedSemanticHost::new(&config("process-1")).unwrap()
}
fn decode(text: &str) -> Value {
    serde_json::from_str(text).unwrap()
}
fn snapshot(state: &TrustedSemanticHost) -> Value {
    decode(&state.snapshot().unwrap())
}
fn current(state: &TrustedSemanticHost, name: &str) -> Value {
    decode(&state.current(name).unwrap())
}
fn operation(user: &str, id: &str) -> Value {
    json!({"userId":user,"clientId":format!("tab-{user}"),"requestId":id})
}
fn change(state: &TrustedSemanticHost, name: &str, before: i32, after: i32) -> Value {
    json!({"address":{"target":current(state,name),"property":"value"},"before":before,"after":after})
}
fn commit(state: &mut TrustedSemanticHost, write: &Value) {
    state
        .commit_stage(write["stageId"].as_str().unwrap())
        .unwrap();
}
fn record(
    state: &mut TrustedSemanticHost,
    user: &str,
    id: &str,
    name: &str,
    before: i32,
    after: i32,
) {
    let basis = snapshot(state)["revision"].as_u64().unwrap();
    let write = decode(&state.stage_validated_record(&json!({"basisRevision":basis,"revision":basis+1,"operation":operation(user,id),"changes":[change(state,name,before,after)]}).to_string()).unwrap());
    commit(state, &write);
}
fn transact(state: &mut TrustedSemanticHost, mut request: Value) -> Value {
    let basis = snapshot(state)["revision"].as_u64().unwrap();
    request["basisRevision"] = json!(basis);
    request["revision"] = json!(basis + 1);
    let write = decode(
        &state
            .stage_validated_transaction(&request.to_string())
            .unwrap(),
    );
    commit(state, &write);
    write
}
fn inverse(state: &mut TrustedSemanticHost, prepared: &Value, user: &str, id: &str) {
    let revision = snapshot(state)["revision"].as_u64().unwrap() + 1;
    let write = decode(
        &state
            .stage_validated_inverse(
                prepared["ticket"].as_str().unwrap(),
                &operation(user, id).to_string(),
                &revision.to_string(),
            )
            .unwrap(),
    );
    commit(state, &write);
}

#[test]
fn initial_inventory_allocates_forward_references_and_rejects_invalid_graph_inputs() {
    let state = host();
    assert_eq!(snapshot(&state)["dependencyCount"], 1);
    let plan = decode(
        &state
            .plan_delete(&json!([current(&state, "width")]).to_string())
            .unwrap(),
    );
    assert_eq!(plan["closure"].as_array().unwrap().len(), 2);
    for objects in [
        json!([{"object":"a","dependencies":["absent"]}]),
        json!([{"object":"a"},{"object":"a"}]),
        json!([{"object":"a","dependencies":["a"]}]),
    ] {
        let configuration =
            json!({"documentEpoch":"doc-1","serverEpoch":"process-1","objects":objects});
        assert!(TrustedSemanticHost::new(&configuration.to_string()).is_err());
    }
    let mut configuration = decode(&config("process-1"));
    configuration["limits"] = json!({"maxObjectsIncludingTombstones":2});
    assert!(TrustedSemanticHost::new(&configuration.to_string()).is_err());
    configuration["limits"] = json!({"maxTotalDependencies":1_000_001});
    assert!(TrustedSemanticHost::new(&configuration.to_string()).is_err());
}

#[test]
fn staging_reads_committed_state_and_rejects_foreign_or_replayed_durability_tokens() {
    let mut state = host();
    let mut other = host();
    let before = state.checkpoint().unwrap();
    let request=json!({"basisRevision":0,"revision":1,"operation":operation("alice","one"),"changes":[change(&state,"width",12,14)]}).to_string();
    let write = decode(&state.stage_validated_record(&request).unwrap());
    let foreign = decode(&other.stage_validated_record(&request).unwrap());
    assert_eq!(state.checkpoint().unwrap(), before);
    assert_eq!(snapshot(&state)["revision"], 0);
    assert!(state.stage_validated_record(&request).is_err());
    assert!(state.prepare_undo("alice").is_err());
    assert!(
        state
            .commit_stage(foreign["stageId"].as_str().unwrap())
            .is_err()
    );
    assert!(
        state
            .fail_stage(foreign["stageId"].as_str().unwrap())
            .is_err()
    );
    assert_eq!(snapshot(&state)["hasPendingStage"], true);
    commit(&mut state, &write);
    commit(&mut other, &foreign);
    assert_eq!(snapshot(&state)["revision"], 1);
    assert_eq!(
        decode(&state.checkpoint().unwrap())["historyJson"],
        write["historyJson"]
    );
    assert!(
        state
            .commit_stage(write["stageId"].as_str().unwrap())
            .is_err()
    );
    assert!(state.stage_validated_record(&request).is_err());
}

#[test]
fn same_value_contributions_preserve_ownership_in_both_actor_orders() {
    for (first, second) in [("alice", "bob"), ("bob", "alice")] {
        let mut state = host();
        record(&mut state, first, "first", "width", 12, 14);
        let old_ticket = decode(&state.prepare_undo(first).unwrap());
        record(&mut state, second, "second", "width", 14, 14);
        assert!(state.prepare_undo(first).unwrap_err().contains("owns"));
        assert!(
            state
                .stage_validated_inverse(
                    old_ticket["ticket"].as_str().unwrap(),
                    &operation(first, "stale").to_string(),
                    "3"
                )
                .is_err()
        );
        let undo = decode(&state.prepare_undo(second).unwrap());
        assert_eq!(undo["changes"][0]["before"], 14);
        assert_eq!(undo["changes"][0]["after"], 14);
        inverse(&mut state, &undo, second, "undo");
        let undo = decode(&state.prepare_undo(first).unwrap());
        assert_eq!(undo["changes"][0]["after"], 12);
        let address = undo["changes"][0]["address"].to_string();
        assert_eq!(
            decode(&state.property_owner(&address).unwrap()),
            operation(first, "first")
        );
    }
}

#[test]
fn stale_delete_cannot_gain_dependents_or_shrink_reviewed_closure_via_same_transaction() {
    let mut state = host();
    record(&mut state, "alice", "width", "width", 12, 14);
    let original = current(&state, "width");
    let plan = decode(&state.plan_delete(&json!([original]).to_string()).unwrap());
    transact(
        &mut state,
        json!({"create":[{"object":"late","dependencies":[{"kind":"existing","target":original}]}]}),
    );
    assert!(state.authenticate_delete(&plan.to_string()).is_err());
    let before = state.checkpoint().unwrap();
    let attempted=json!({"basisRevision":2,"revision":3,"deletions":[plan],"dependencies":[{"target":{"kind":"existing","target":current(&state,"late")},"dependencies":[]}],"create":[{"object":"should-not-exist"}]}).to_string();
    assert!(state.stage_validated_transaction(&attempted).is_err());
    assert_eq!(state.checkpoint().unwrap(), before);
    let current_plan = decode(&state.plan_delete(&json!([original]).to_string()).unwrap());
    let write = transact(
        &mut state,
        json!({"deletions":[current_plan],"create":[{"object":"width"},{"object":"child","dependencies":[{"kind":"created","object":"parent"}]},{"object":"parent","dependencies":[{"kind":"created","object":"width"}]}]}),
    );
    assert_eq!(write["created"].as_array().unwrap().len(), 3);
    assert!(
        current(&state, "width")["generation"].as_u64().unwrap()
            > original["generation"].as_u64().unwrap()
    );
    assert!(state.authenticate(&original.to_string()).is_err());
    assert!(
        state
            .prepare_undo("alice")
            .unwrap_err()
            .contains("recreated")
    );
    assert_eq!(current(&state, "edge"), Value::Null);
    assert_eq!(current(&state, "late"), Value::Null);
    assert_eq!(snapshot(&state)["historyRevision"], 1);
    assert_eq!(snapshot(&state)["revision"], 3);
}

#[test]
fn lifecycle_and_property_batch_fail_atomically_then_commit_together() {
    let mut state = host();
    record(&mut state, "alice", "first", "width", 12, 14);
    let before = state.checkpoint().unwrap();
    let mut request = json!({"basisRevision":1,"revision":2,"create":[{"object":"new"}],"record":{"operation":operation("alice","second"),"changes":[change(&state,"width",12,16)]}});
    assert!(
        state
            .stage_validated_transaction(&request.to_string())
            .is_err()
    );
    assert_eq!(state.checkpoint().unwrap(), before);
    assert_eq!(current(&state, "new"), Value::Null);
    request["record"]["changes"][0]["before"] = json!(14);
    let write = decode(
        &state
            .stage_validated_transaction(&request.to_string())
            .unwrap(),
    );
    assert_eq!(current(&state, "new"), Value::Null);
    commit(&mut state, &write);
    assert_ne!(current(&state, "new"), Value::Null);
    assert_eq!(snapshot(&state)["historyRevision"], 2);
    let request=json!({"basisRevision":2,"revision":3,"dependencies":[{"target":{"kind":"created","object":"new"},"dependencies":[]}]}).to_string();
    assert!(state.stage_validated_transaction(&request).is_err());
}

#[test]
fn restart_preserves_exact_history_redo_and_rejects_foreign_checkpoints_and_tickets() {
    let mut state = host();
    record(&mut state, "alice", "width", "width", 12, 14);
    record(&mut state, "bob", "height", "height", 8, 9);
    let undo = decode(&state.prepare_undo("alice").unwrap());
    inverse(&mut state, &undo, "alice", "undo");
    let old = decode(&state.prepare_redo("alice").unwrap());
    let checkpoint = state.checkpoint().unwrap();
    let mut restored = TrustedSemanticHost::restore(&config("new-process"), &checkpoint).unwrap();
    assert_eq!(restored.checkpoint().unwrap(), checkpoint);
    assert!(
        restored
            .stage_validated_inverse(
                old["ticket"].as_str().unwrap(),
                &operation("alice", "redo").to_string(),
                "4"
            )
            .is_err()
    );
    let redo = decode(&restored.prepare_redo("alice").unwrap());
    inverse(&mut restored, &redo, "alice", "redo");
    assert_eq!(snapshot(&restored)["revision"], 4);
    let address = change(&restored, "height", 8, 9)["address"].to_string();
    assert_eq!(
        decode(&restored.property_owner(&address).unwrap()),
        operation("bob", "height")
    );
    let mut foreign = decode(&checkpoint);
    foreign["documentEpoch"] = json!("other");
    assert!(TrustedSemanticHost::restore(&config("new"), &foreign.to_string()).is_err());
    let mut ahead = decode(&checkpoint);
    ahead["revision"] = json!(2);
    assert!(TrustedSemanticHost::restore(&config("new"), &ahead.to_string()).is_err());
    let mut corrupt = decode(&checkpoint);
    let mut history = decode(corrupt["historyJson"].as_str().unwrap());
    history["properties"] = json!([]);
    corrupt["historyJson"] = json!(history.to_string());
    assert!(TrustedSemanticHost::restore(&config("new"), &corrupt.to_string()).is_err());
}

#[test]
fn uncertain_persistence_requires_recovery_and_inverse_handles_are_bounded() {
    let mut state = host();
    record(&mut state, "alice", "first", "width", 12, 14);
    let mut handles = Vec::new();
    for _ in 0..32 {
        handles.push(decode(&state.prepare_undo("alice").unwrap()));
    }
    assert!(
        state
            .prepare_undo("alice")
            .unwrap_err()
            .contains("capacity")
    );
    state
        .release(handles[0]["ticket"].as_str().unwrap())
        .unwrap();
    assert!(state.prepare_undo("alice").is_ok());
    let write = decode(
        &state
            .stage_validated_inverse(
                handles[1]["ticket"].as_str().unwrap(),
                &operation("alice", "undo").to_string(),
                "2",
            )
            .unwrap(),
    );
    let durable=json!({"documentEpoch":"doc-1","revision":2,"targetsJson":write["targetsJson"],"historyJson":write["historyJson"]}).to_string();
    state
        .fail_stage(write["stageId"].as_str().unwrap())
        .unwrap();
    assert_eq!(snapshot(&state)["revision"], 1);
    assert_eq!(snapshot(&state)["needsRecovery"], true);
    assert!(
        state
            .prepare_undo("alice")
            .unwrap_err()
            .contains("recovery")
    );
    assert!(
        state
            .commit_stage(write["stageId"].as_str().unwrap())
            .is_err()
    );
    let mut restored = TrustedSemanticHost::restore(&config("new"), &durable).unwrap();
    assert_eq!(snapshot(&restored)["revision"], 2);
    assert!(restored.prepare_redo("alice").is_ok());
}
